src/iceberg/file_io.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "iceberg/error.h"
+#include "iceberg/expected.h"
+#include "iceberg/iceberg_export.h"
+
+namespace iceberg {
+
+/// \brief Pluggable module for reading, writing, and deleting metadata files.
+///
+/// This module only handle metadata files, not data files. The metadata files
+/// are typically small and are used to store schema, partition information,
+/// and other metadata about the table.
+class ICEBERG_EXPORT FileIO {
+ public:
+  FileIO() = default;
+  virtual ~FileIO() = default;
+
+  /// \brief Read the content of the file at the given location.
+  ///
+  /// \param file_location The location of the file to read.
+  /// \param length The number of bytes to read. Some object storage need to specify
+  /// the length to read, e.g. S3 `GetObject` has a Range parameter.
+  /// \return The content of the file if the read succeeded, an error code if the read
+  /// failed.
+  virtual expected<std::string, Error> ReadFile(const std::string& file_location,
+                                                std::optional<size_t> length) {
+    // The following line is to avoid Windows linker error LNK2019.
+    // If this function is defined as pure virtual function, the `unexpected<Error>` will
+    // not be instantiated and exported in libiceberg.
+    return unexpected<Error>{
```

Comment:
Why not leaving them pure virtual just like `DeleteFile`?

Reply:
Then the iceberg::expected<void,struct iceberg::Error> will not be instantiated on windows platform[0]. I will change the comment to clarify this.

[0] https://github.com/zhjwpku/iceberg-cpp/actions/runs/14125045321/job/39572213082

Reply:
Then provide a default impl for `DeleteFile` to be consistent?

Reply:
done as suggested.